[Cleon2] iP

Cleon2 opened the pull request:
# Duke

## Description

This iP PR aims to enhance our project by implementing several key features and addressing important issues. The following lists outline the changes and tasks to be completed:

### Features Added:

- [x] 🌟 Added ability to add and delete tasks
- [x] 📊 Added validation
- [x] 🛡️ Added list functionality
- [x] 🔄 Integrated persistent storage

### Issues Resolved:

1. 🐛 Fixed a critical bug causing data to fail to persist
2. 🐞 Addressed performance issues related to database queries.

### Code Changes:

```java
# Example Java code snippet with syntax highlighting
package duke;

import duke.commands.Command;
import duke.exceptions.DukeException;
import duke.parser.Parser;
import duke.storage.Storage;
import duke.tasks.DukeList;
import duke.ui.Ui;

/**
 * Represents the main application class for Duke.
 */
public class Duke {

    private DukeList ItemList;
    private Storage storage;
    private Ui ui;

    /**
     * Constructs a Duke instance with the given file path.
     *
     * @param filePath The path to the data file.
     */
    public Duke(String filePath) {
        ui = new Ui();
        this.storage = new Storage(filePath);

        try {
            this.ItemList = new DukeList(this.storage.load());
        } catch (DukeException e) {
            ui.showLoadingError();
            this.ItemList = new DukeList();
        }
    }

    /**
     * The main method that initializes and runs the Duke application.
     *
     * @param args Command line arguments.
     */
    public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }

    /**
     * Runs the Duke application.
     */
    public void run() {
        // Initializing Scanner
        ui.showWelcome();
        boolean isExit = false;
        while (!isExit) {
            try {
                String fullCommand = ui.readCommand();
                ui.showLine();
                Command c = Parser.parse(fullCommand);
                c.execute(ItemList, ui, storage);
                isExit = c.isExit();
            } catch (DukeException e) {
                ui.showError(e.getMessage());
            } finally {
                ui.showLine();
            }
        }
    }
}

```

### Task List:
- * Conduct a security audit to identify potential vulnerabilities.
- * Update documentation with the latest changes.
- * Prepare a release candidate for deployment.

### Additional Notes:

> 
For more details, please refer to the [full PR here](https://github.com/Cleon2/ip).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Cleon2 committed:
Level-0

Cleon2 committed:
Level-1

Cleon2 committed:
Level-3

Cleon2 committed:
Level-4

Cleon2 committed:
Level-5

Cleon2 committed:
Level-6

Cleon2 committed:
A-TextUiTesting

Cleon2 committed:
Level-7

Cleon2 committed:
Merge branch 'branch-Level-7'

Cleon2 committed:
Level-8

Cleon2 committed:
merge branch 'branch-Level-8'

Cleon2 committed:
A-MoreOOP

Cleon2 committed:
Merge branch 'A-MoreOOP'

Cleon2 committed:
A-Packages

Cleon2 committed:
Merge branch 'branch-Packages'

Cleon2 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip

Cleon2 committed:
A-Gradle

Cleon2 committed:
A-JUnit

Cleon2 committed:
A-JavaDoc

Cleon2 committed:
A-CodingStandard

Cleon2 committed:
Level-9

Cleon2 committed:
Merge branch 'branch-A-JavaDoc'

Cleon2 committed:
resolved conflict

Cleon2 committed:
Level-9

Cleon2 committed:
Level-9

Cleon2 committed:
added acknowledgements

Cleon2 committed:
fixed coding violations

Cleon2 committed:
Merge branch 'branch-A-CheckStyle'

Cleon2 committed:
added GUI

Cleon2 committed:
Merge branch 'branch-Level-10'

Cleon2 committed:
added gui for real

Cleon2 committed:
Merge branch 'branch-Level-10'

kayabuttertoastt commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public Duke() {
+        ui = new Ui();
+        ui = new Ui();
```
Any reason why 'ui' is initialised twice instead of once?

kayabuttertoastt commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Represents an abstract command that can be executed.
+ */
+public abstract class Command {
```
I like the nice usage of inheritance here which makes it much more extensible! :)

kayabuttertoastt commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,52 @@
+     */
+    @Override
+    public String execute(DukeList tasks, Ui ui, Storage storage) throws DukeException {
+        if (tasks.getSize() < index) {
```
Perhaps it would be better to also consider handling errors for the situation when the user inputs an index greater than `tasks.getSize()`.

kayabuttertoastt commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,222 @@
+ * Represents a parser to convert user input into executable commands.
+ */
+//Solution below inspired by Shishir Bychapur
+public class Parser {
```
I like your thorough approach to identifying invalid inputs and delivering pertinent and specific error messages to the user.

kayabuttertoastt reviewed (commented):
Overall, LGTM, just a few nits.

nicrandomlee commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Represents a custom exception specific to the Duke application.
+ */
+public class DukeException extends Exception {
```
Good use of your own custom exception class 👍

nicrandomlee commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,222 @@
+        String[] split = input.split(" ", 2);
+        Command c = null;
+
+        switch (split[0]) {
```
Good use of switch statement to structure code logically 👍

nicrandomlee commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,121 @@
+        try {
+            String[] splitted = path.split("/");
+            File directory = new File(splitted[0]);
+            if (!directory.exists()) {
```
Good handling of edge cases

nicrandomlee reviewed (commented):
Overall, your code is very well written and follows code quality standards. I am very thankful to have reviewed your code and I have learnt a lot from you. Keep it up 👍

Cleon2 committed:
Remove unnecessary instantiation of ui in duke class

ui has already been instantiated

let's remove that unnecessary line

:

Cleon2 committed:
Add assertions

Empty task descriptions may be entered and invalid deletion indexes might be keyed

Add assertions to check for empty descriptions or invalid indexes

Cleon2 committed:
Merge pull request #1 from Cleon2/branch-A-Assertions

Add assertions

Cleon2 committed:
fixed code quality

Cleon2 committed:
Merge pull request #2 from Cleon2/branch-A-CodeQuality

fixed code quality

Cleon2 committed:
C-BetterSearch

Cleon2 committed:
Merge pull request #3 from Cleon2/branch-C-BetterSearch

C-BetterSearch

Cleon2 committed:
add a user guide

Cleon2 committed:
Merge pull request #5 from Cleon2/branch-A-UserGuide

add a user guide
